### PR TITLE
feat(subscriptions): Add `TaskSetConsumer`

### PR DESCRIPTION
**Pull request by tkaemming**
This change adds a `TaskSetConsumer` class that implements an API that is similar to the `Consumer` API, returning a `TaskSet` instance instead of a `Message`.

Additional implementation details are in the `TaskSetConsumer` docstring and implementation comment.

### CODE REVIEW

File: snuba/subscriptions/consumer.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+            if on_revoke is not None:
+                on_revoke(assignment)
+
+        self.__consumer.subscribe(
+            topics, on_assign=on_assign_callback, on_revoke=on_revoke_callback
+        )
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[TaskSet]:
+        message = self.__consumer.poll(*[timeout] if timeout is not None else [])
```
Comment (tkaemming) [inline snuba/subscriptions/consumer.py]:
```
>>> from confluent_kafka import Consumer
>>> Consumer({}).poll(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: must be real number, not NoneType
```

🤷‍♂

File: snuba/subscriptions/consumer.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        Return a new ``Interval`` instance, using the provided position as
+        the new upper bound, and the existing upper bound as the new lower
+        bound.
+        """
+        return Interval(self.upper, position)
+
+
+@dataclass(frozen=True)
+class Task:
+    """Represents a single task to be executed."""
```
Comment (tkaemming) [inline snuba/subscriptions/consumer.py]: Details within this class are to be filled in, but ultimately this will have some mechanism for building a `Query` object to be executed.

File: snuba/subscriptions/consumer.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    value = configuration.get(key, default)
+    if isinstance(value, str):
+        return value.lower().strip() == "true"
+    elif isinstance(value, bool):
+        return value
+    else:
+        raise TypeError(f"unexpected value for {key!r}")
+
+
+class PartitionState:
```
Comment (tkaemming) [inline snuba/subscriptions/consumer.py]: I'm not sold on this abstraction yet, but also trying to keep it simple for now.

File: snuba/subscriptions/consumer.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+                    self.__topic_partition_states[key] = state
+                    logger.debug("Initialized partition state for %r: %r", key, state)
+
+            if on_assign is not None:
+                on_assign(assignment)
+
+        def on_revoke_callback(
+            consumer: Any, assignment: Sequence[TopicPartition]
+        ) -> None:
+            # XXX: Check to see if this causes any weird shit during rebalancing.
```
Comment (tkaemming) [inline snuba/subscriptions/consumer.py]: Testing this is going to be a challenge.

File: snuba/subscriptions/consumer.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        state = self.__topic_partition_states[
+            (message.topic(), message.partition())
+        ].send(Position(message.offset(), timestamp / 1000.0))
+
+        if isinstance(state, Position):
+            tasks = None
+        elif isinstance(state, Interval):
+            # TODO: Fetch tasks that are scheduled between the interval
+            # endpoints for this topic and partition.
+            tasks = TaskSet(message.topic(), message.partition(), state, [])
```
Comment (tkaemming) [inline snuba/subscriptions/consumer.py]: There will be a separate consumer that consumes subscription state (either in another thread, or in the same consumer with pausing and resuming partitions after subscription state is initialized) and returns the `Sequence[Task]` here.

File: snuba/subscriptions/consumer.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+            self.__state = Interval(self.__state, position)
+            return self.__state
+        elif isinstance(self.__state, Interval):
+            self.__state = self.__state.shift(position)
+            return self.__state
+        else:
+            raise ValueError("unexpected state")
+
+
+class TaskSetConsumer(Consumer[TaskSet]):
```
Comment (tkaemming) [inline snuba/subscriptions/consumer.py]: See note in `snuba/utils/kafka/consumer.py` about this for context.

File: snuba/utils/kafka/consumer.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Generic, Optional, Sequence, TypeVar
+
+from confluent_kafka import TopicPartition
+
+
+TMessage = TypeVar("TMessage")
+
+
+class Consumer(Generic[TMessage], ABC):
```
Comment (tkaemming) [inline snuba/utils/kafka/consumer.py]: This is (roughly) the API that is used by the `BatchingKafkaConsumer`.

My thinking here: I would like to avoid introducing another way of running Kafka consumers to this project (and reimplementing things like commit batching, etc.)

To effectively support subscriptions, I think it will essentially require adding streaming support to the `BatchingKafkaConsumer` library, since I think it's important to be able to execute each `TaskSet` as soon as possible without waiting for a batch size limit or timeout. I do think it's useful (and important) to keep the same batch *commit* semantics, though (and configuration consistency.)

This will require some relatively substantial changes to the library (and ones that we've been talking about for a while with respect to streaming writes for memory usage.) My thinking is that we should vendor the `BatchingKafkaConsumer` into Snuba (at least temporarily) to avoid having to synchronize these changes across multiple projects.

Long-term, I'm not sure how we want to manage this dependency, as well as the `SynchronizedConsumer` dependency that will be introduced. Curious on everyone else's thoughts here.

File: snuba/subscriptions/consumer.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+        # Using the built-in ``enable.auto.offset.store`` behavior will cause
+        # gaps in task execution during restarts or rebalances, since it will
+        # set the next offset to be consumed as the most recently message
+        # consumed's offset + 1. In our case, we always want the first message
+        # to be consumed after a rebalance to be *the same* as the previously
+        # consumed message. To account for this, we have to always disable it,
+        # and use our own implementation that stores the offset of the most
+        # recently received message in a partition, instead of that offset + 1.
+        self.__consumer = KafkaConsumer(
```
Comment (tkaemming) [inline snuba/subscriptions/consumer.py]: This will eventually need to act similarly to the `SynchronizedConsumer`, since we shouldn't be executing queries that haven't been written yet. That does mean we'll have to extract the `SynchronizedConsumer` from Sentry at some point (either by copying, or putting it in a library.)

**Review by tkaemming — COMMENTED** [review]
I've experimented with several different approaches to managing subscription execution, and this is the least worst approach I've come up with so far. It will take some additional work before its ready to merge, but I wanted to get some feedback before investing more time in case anybody has better ideas or concerns with this approach.

Other than tests, I'd like this to be in a point where the consumer can be invoked via the CLI (not necessarily actually executing tasks, or doing any sort of consumer synchronization) before I feel comfortable enough with this to consider it ready to merge.

File: snuba/utils/kafka/consumer.py (hunk 7 above)
Comment (tkaemming) [reply snuba/utils/kafka/consumer.py]: > Long-term, I'm not sure how we want to manage this dependency, as well as the `SynchronizedConsumer` dependency that will be introduced. Curious on everyone else's thoughts here.

I also think it's relatively important to have this code type checked. I haven't looked into how mypy works with third party modules that need to support Python 2 ([using comment-based syntax](https://mypy.readthedocs.io/en/latest/python2.html)) and 3.

Comment (tkaemming) [reply snuba/utils/kafka/consumer.py]: This would also tighten up the specification of the [`FakeKafkaConsumer`](https://github.com/getsentry/snuba/blob/d4fc8330b8e6abcd2d350f794933f6794975cf2f/tests/base.py#L76-L117) for testing, which I think is a good thing.

**Review by fpacifici — COMMENTED** [review]
Mostly questions at this stage, but I like the approach.

There is a use case during failover I am not sure I follow.
Let's say I do this:
consume MA
store MA offset
consumer MB
commit MB
consume MC
save offset, do not commit
restart and the broker commits MC because enough time has passed
consume MD 
consume ME
return tasks between MD-ME

I would never execute the tasks between MC and MD. Is this the case and is this expected?

**Comment by fpacifici** [comment]
I find it convenient to use tests to document some logic like the one on this class that has corner cases that are not evident (like https://github.com/getsentry/snuba/blob/master/tests/stateful_consumer/test_bootstrap_state.py).
I think something like that would help as soon as you have some logic that returns tasks

File: snuba/subscriptions/consumer.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+        ].send(Position(message.offset(), timestamp / 1000.0))
+
+        if isinstance(state, Position):
+            tasks = None
+        elif isinstance(state, Interval):
+            # TODO: Fetch tasks that are scheduled between the interval
+            # endpoints for this topic and partition.
+            tasks = TaskSet(message.topic(), message.partition(), state, [])
+        else:
+            raise ValueError("unexpected state")
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: I think pushing this into the state object is a good idea, going to follow up with a change that addresses this.

File: snuba/subscriptions/consumer.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+                offsets=[
+                    TopicPartition(
+                        message.topic(), message.partition(), message.offset()
+                    )
+                ]
+            )
+
+        return tasks
+
+    def commit(self, asynchronous: bool = True) -> Optional[Sequence[TopicPartition]]:
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: It's required by the batching consumer implementation, and it shouldn't break that invariant — hopefully the earlier commit makes it clearer how this works.

File: snuba/subscriptions/consumer.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+        elif isinstance(state, Interval):
+            # TODO: Fetch tasks that are scheduled between the interval
+            # endpoints for this topic and partition.
+            tasks = TaskSet(message.topic(), message.partition(), state, [])
+        else:
+            raise ValueError("unexpected state")
+
+        if self.__enable_auto_offset_store:
+            self.__consumer.store_offsets(
+                offsets=[
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: This should work as expected, and I think there is probably some confusion due to the difference in the way that the Kafka driver handles offset storage and committed message offsets (which I perpetuated in my long comment that should probably be rewritten to be clearer.)

The important thing to note is that committed offset for a partition is the first offset that the a consumer will receive from that partition after restarting or rebalancing. This is different than committing a _message_, which is to commit that message's offset + 1 so that that same message isn't fetched again by a consumer in that partition.

This example using [this test code](https://gist.github.com/tkaemming/6ed94da5ae9dd4b352ff626006d24584 ) should demonstrate this behavior more clearly (I deleted some logging to make it a bit easier to understand):

```python
>>> produce('example-topic')
>>> consume('example-topic')
no message
example-topic 0 0 b'0'
example-topic 0 1 b'1'
example-topic 0 2 b'2'
example-topic 0 3 b'3'
example-topic 0 4 b'4'
no message
commit [TopicPartition{topic=example-topic,partition=0,offset=4,error=None}]
no message
^C
```

At this point, messages 0 through 3 (inclusive) have been committed, message 4 has not been committed, but the committed offset is the same as the offset of message 4. Restarting the consumer will then resume with message 4:

```python
>>> consume('example-topic', 5)
no message
example-topic 0 4 b'4'
example-topic 0 5 b'5'
example-topic 0 6 b'6'
example-topic 0 7 b'7'
example-topic 0 8 b'8'
no message
^C
```

So, using the Interval(MA, MB) example again on cold start:

- First call to `poll`: the consumer receives MA, the consumer stores the offset of MA, the consumer returns None.
  - If at this point we either commit offsets or crash, we'll restart exactly where we left off: the first message consumed will again be MA.
- First call to `poll`: the consumer receives MB, the consumer stores the offset of MB, the consumer returns the task set for Interval(MA, MB).
  - If we do not have auto commit (`enable.auto.commit`) enabled and crash here, we'll restart with MA.
  - If we do not have auto commit enabled and we explicitly commit the stored offset (e.g. after the tasks are completed) and crash here, we'll restart with MB.
  - If we _do_ have auto commit enabled and crash here, one of those two things will happen indeterminately.

File: snuba/utils/kafka/consumer.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
+from abc import ABC, abstractmethod
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/utils/kafka/consumer.py]: I'd prefer there to be "kafka" in the name, and keep `snuba.consumers` for higher-level abstract consumers, personally. I'm not sure `utils` is a great place to put it but it seems better than as a top level thing.

File: snuba/subscriptions/consumer.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+        elif isinstance(self.__state, Interval):
+            self.__state = self.__state.shift(position)
+            return self.__state
+        else:
+            raise ValueError("unexpected state")
+
+
+class TaskSetConsumer(Consumer[TaskSet]):
+    """
+    Identifies tasks to be executed based on the data within a Kafka topic.
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Makes sense, will do.

File: snuba/subscriptions/consumer.py [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+    """
+
+    def __init__(self, configuration: Mapping[str, Any]) -> None:
+        # There are three valid states for a partition in this mapping:
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: I'll put some thought into this — I don't think it should be moved to `PartitionState` wholesale because this deals with offset commit semantics which I'd like to keep out of that class. I'll try and split this up when addressing some of the partition state comments below.

**Review by tkaemming — COMMENTED** [review]
> I would never execute the tasks between MC and MD. Is this the case and is this expected?

I don't think this is an accurate read. Specifically, I don't see how this would happen:

> restart and the broker commits MC because enough time has passed

Offset storage is a consumer-local thing and is just used for staging offsets to be committed. It's true that if using auto commit, offsets could be preemptively committed, but I don't see us actually using that anywhere. Otherwise, I think this is generally addressed more specifically be the long comment inline.

File: snuba/subscriptions/consumer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/subscriptions/consumer.py]: ok, understood. Makes sense.
Though my followup is whether the behavior is the same if someone commits after receiving a task list by calling the commit method. In this case you are specifying the offset, thus you are right, you will restart from that offset you saved. But in the commit case you are not specifying an offset. The documentation says

```
message and offsets are mutually exclusive, if neither is set the current partition assignment’s offsets are used instead
```

So, in this case what is "the current partition assignment" ? Is it the offset of the last consumed message + 1 or is it the last offset you stored with store_offset ?
If calling commit after getting the taskSet between MA and MB commits MB+1, that would be the upper bound. But if it is the last offset you stored then it works.

[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: > So, in this case what is "the current partition assignment" ? Is it the offset of the last consumed message + 1 or is it the last offset you stored with store_offset ?

My understanding is that it should be the latter (the last offset you stored with `store_offset`), but I'll double check.

[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Ok, I just reran the same script with autocommit disabled and an explicit synchronous commit in the `except KeyboardInterrupt` clause:

```diff
diff --git a/test.py b/test.py
index 7fcdd1b..cdf3199 100644
--- a/test.py
+++ b/test.py
@@ -22,7 +22,7 @@ def consume(topic, n=5):
             {
                 "group.id": "test",
                 "auto.offset.reset": "earliest",
-                "enable.auto.commit": "true",
+                "enable.auto.commit": "false",
                 "enable.auto.offset.store": "false",
                 "default.topic.config": {"auto.offset.reset": "earliest"},
                 "on_commit": on_commit,
@@ -56,4 +56,5 @@ def consume(topic, n=5):
             if i >= n:
                 consumer.pause([TopicPartition(message.topic(), message.partition())])
     except KeyboardInterrupt:
+        consumer.commit(asynchronous=False)
         consumer.close()
```

It worked as expected (except apparently that doesn't trigger the `commit_cb`):

```python
>>> consume('example')
no message
example 0 10 b'0'
example 0 11 b'1'
example 0 12 b'2'
example 0 13 b'3'
example 0 14 b'4'
no message
^C
```

```python
>>> consume('example')
no message
example 0 14 b'4'
example 0 15 b'5'
example 0 16 b'6'
example 0 17 b'7'
example 0 18 b'8'
^C
```

[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Also, per the documentation: 

> The stored offsets will be committed according to ‘auto.commit.interval.ms’ or manual offset-less commit().

File: snuba/subscriptions/consumer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: I changed the return type of `send` (now `set_position`) to only return an `Interval` publically: https://github.com/getsentry/snuba/blob/349f44494607f6fc36895c313cb1f944eb9cb1b6/snuba/subscriptions/consumer.py#L105-L110

File: snuba/subscriptions/consumer.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: I added a docstring containing the information that I think would be useful to anyone using this abstraction, as well as reworded the comment so that it's clearer (I think) about how offset management is impacted by dealing with intervals rather than messages: https://github.com/getsentry/snuba/blob/349f44494607f6fc36895c313cb1f944eb9cb1b6/snuba/subscriptions/consumer.py#L124-L198

File: snuba/subscriptions/consumer.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: I kind of split this up, but I think the `PartitionState` documentation could still use some work.

File: snuba/subscriptions/consumer.py [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+                    TopicPartition(
+                        message.topic(), message.partition(), message.offset()
+                    )
+                ]
+            )
+
+        return tasks
+
+    def commit(self, asynchronous: bool = True) -> Optional[Sequence[TopicPartition]]:
+        offsets: Optional[Sequence[TopicPartition]] = self.__consumer.commit(
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: It will throw when called synchronously, or include a Kafka error to the callback when called asynchronously.

File: snuba/subscriptions/consumer.py [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+        error = message.error()
+        if error is not None:
+            raise error
+
+        timestamp_type, timestamp = message.timestamp()
+        assert timestamp_type == TIMESTAMP_LOG_APPEND_TIME
+
+        state = self.__topic_partition_states[
+            (message.topic(), message.partition())
+        ].send(Position(message.offset(), timestamp / 1000.0))
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: This is called `set_position` now, which seems better (more specific than `update_state`, I think.)

File: snuba/subscriptions/consumer.py [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+                        f'unexpected value for "auto.offset.reset": {self.__auto_offset_reset!r}'
+                    )
+
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[TopicPartition]], Any]] = None,
+        on_revoke: Optional[Callable[[Sequence[TopicPartition]], Any]] = None,
+    ) -> None:
+        def on_assign_callback(
```
Comment (fpacifici) [inline snuba/subscriptions/consumer.py]: I'd consider moving the part of the comment that explains the rebalance problem here (as docstring of the subscribe method). It would be more contextual and reduce the size of that wall of text.

File: snuba/subscriptions/consumer.py [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+    the committed offset for that partition (or if no offsets have been
+    committed for that partition, the offset is reset in accordance with the
+    ``auto.offset.reset`` configuration value.) This causes partitions that
+    are maintained across a rebalance to have the same offset management
+    behavior as a partition that is moved from one consumer to another. To
+    prevent uncommitted messages from being consumed multiple times,
+    ``commit`` should be called in the partition revocation callback.
+
+    This consumer can either be used with automatic timed offset commit (the
+    librdkafka default, set using the ``enable.auto.commit`` configuration
```
Comment (fpacifici) [inline snuba/subscriptions/consumer.py]: I assume this means that enable.auto.commit has, upon restart, the same behavior than if we consumer manually:
consume offset 5
commit offset 5
restart
consumer offset 5

Is this correct?

**Review by fpacifici — COMMENTED** [review]
The logic is solid. I just have a few cosmetic suggestions. Will let you wrap it up and remove the "draft" flag

File: snuba/subscriptions/consumer.py (hunk 2 above)
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Makes sense, will do.

File: snuba/subscriptions/consumer.py (hunk 18 above)
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Yes. Is this worth calling out specifically?

File: snuba/subscriptions/consumer.py [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+    value) or manual calls to ``commit``.
+    """
+
+    # Since this consumer deals with time intervals between messages rather
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: I wanted to split this in a way where the docstring is the things that you should know if you are using the consumer, and the comment is things you should know if you're editing the consumer itself. There is a little bit of redundancy but I think that is okay (and I think with the subtle complexities here it's better to err on the side of being too thorough rather than leaving out detail.) With that in mind, are there still changes that you think make sense here? I'm open to make some, I think I'm just too entrenched in the details to see what would make sense to change here.

File: snuba/subscriptions/consumer.py [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+    # a ``poll`` call. To avoid this, we reset *every* partition to its
+    # committed offset on assignment (even partitions that were previously
+    # owned by this consumer.) This can lead to duplicated delivery of messages
+    # that were consumed but not committed by this consumer -- but that would
+    # be the same outcome if this partition was moved between independent
+    # consumers anyway.
+
+    def __init__(self, configuration: Mapping[str, Any]) -> None:
+        self.__topic_partition_states: MutableMapping[
+            Tuple[str, int], PartitionState
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: I was on the fence about making a `PartitionKey` namedtuple or something, will go ahead and do that.

File: snuba/subscriptions/consumer.py (hunk 17 above)
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Interesting point, I'll see if I can make that work.

File: snuba/utils/kafka/consumer.py (hunk 7 above)
Comment (tkaemming) [reply snuba/utils/kafka/consumer.py]: This is happening in #485 (you're probably aware of this, mostly noting for posterity.)

File: snuba/subscriptions/consumer.py (hunk 2 above)
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Fixed with 6f7962ea0d95838717caff4167b06074cf703027.

File: snuba/subscriptions/consumer.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Added `PartitionKey` with bd59d0c7f1202a89e9809da47ee731cc3575aa6b.

File: snuba/subscriptions/consumer.py (hunk 17 above)
Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Moved and slightly reworded that comment with aaef1c6c3e648531132743f695a6d14e80baf9e4.

Comment (tkaemming) [reply snuba/subscriptions/consumer.py]: Moved and slightly reworded that comment with aaef1c6c3e648531132743f695a6d14e80baf9e4.

**Comment by tkaemming** [comment]
Trying to rebase this would be a bit of a nightmare — going to close this and reopen as a separate PR on top of the changes introduced in #512.
